modules/mauth-authenticator-http4s/src/test/scala/com/mdsol/mauth/http4s/RequestAuthenticatorSpec.scala

```diff
@@ -0,0 +1,151 @@
+package com.mdsol.mauth.http4s
+
+import cats.effect.IO
+import com.mdsol.mauth.RequestAuthenticatorBaseSpec
+import com.mdsol.mauth.exception.MAuthValidationException
+import com.mdsol.mauth.scaladsl.utils.ClientPublicKeyProvider
+import com.mdsol.mauth.test.utils.FakeMAuthServer.EXISTING_CLIENT_APP_UUID
+import com.mdsol.mauth.util.{EpochTimeProvider, MAuthKeysHelper}
+import munit.CatsEffectSuite
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+
+import java.security.Security
+import java.util.UUID
+import scala.concurrent.duration._
+
+class RequestAuthenticatorSpec extends CatsEffectSuite with RequestAuthenticatorBaseSpec {
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    Security.addProvider(new BouncyCastleProvider)
+    ()
+  }
+
+  private implicit val requestValidationTimeout: Duration = 10.seconds
+
+  def mockEpochTimeProvider(seconds: Long): EpochTimeProvider = () => seconds
+
+  private val client: ClientPublicKeyProvider[IO] = (appUUID: UUID) =>
+    if (appUUID == EXISTING_CLIENT_APP_UUID) IO.pure(Some(MAuthKeysHelper.getPublicKeyFromString(PUBLIC_KEY)))
+    else IO.none
+
+  test("authenticate a valid request") {
+    //noinspection ConvertibleToMethodValue
+    val authenticator = RequestAuthenticator[IO](client, mockEpochTimeProvider(CLIENT_X_MWS_TIME_HEADER_VALUE.toLong + 3))
+
+    authenticator.authenticate(getSimpleRequest).assert
+  }
+
+  test("authenticate a request with unicode chars in body") {
+    val authenticator = RequestAuthenticator(client, mockEpochTimeProvider(CLIENT_UNICODE_X_MWS_TIME_HEADER_VALUE.toLong + 3))
+
+    authenticator.authenticate(getRequestWithUnicodeCharactersInBody).assert
+  }
+
+  test("authenticate a request without any body") {
+    val authenticator = RequestAuthenticator(client, mockEpochTimeProvider(CLIENT_NO_BODY_X_MWS_TIME_HEADER_VALUE.toLong + 3))
+
+    authenticator.authenticate(getRequestWithoutMessageBody).assert
+  }
+
+  test("not authenticate an invalid request") {
+    val authenticator = RequestAuthenticator(client, mockEpochTimeProvider(CLIENT_X_MWS_TIME_HEADER_VALUE.toLong + 3))
+    val invalidRequest = getSimpleRequestWithWrongSignature
+    authenticator.authenticate(invalidRequest).assertEquals(false)
+  }
+
+  test("not authenticate a request after timeout period passed") {
+    val client: ClientPublicKeyProvider[IO] = _ => IO.none
+
+    val authenticator = RequestAuthenticator(client, mockEpochTimeProvider(CLIENT_UNICODE_X_MWS_TIME_HEADER_VALUE.toLong + 600))
+
+    authenticator.authenticate(getRequestWithUnicodeCharactersInBody).handleErrorWith {
+      case e: MAuthValidationException =>
+        IO.pure(assertEquals(e.getMessage, "MAuth request validation failed because of timeout 10 seconds"))
+      case _ => fail("should not be here")
+    }
```

Comment:
unfortunately this doesn't work as it would seem. The `assertEquals` inside the `IO` is never executed.
The correct way to prove that this works correctly is to call 
`authenticator.authenticate(getRequestWithUnicodeCharactersInBody).attempt` and deal with the disjoint union.
`assert` / `assertEqual` has to be the last instruction evaluated in the test scenario in order to get a valid test result (here instead it's nested within the `IO`, to clarify what I mean)

Reply:
Oh thats fun. It would work in out IOWordSpec scalatest suite :D

Reply:
Yeah, that's a quirk of MUnit. It's really easy to overlook that.

Reply:
MUnit had a useful intercept that works a treat and it looks cleaner tbh